### PR TITLE
chore: split `config.Registries` into the separate resource.

### LINKED ISSUE #9766: image cache: split out config.Registries to a resource and re-wire the controllers

### PR DESCRIPTION
**Pull request by DmitriyMV**
Required for #9614

Closes #9766

### CODE REVIEW

File: pkg/machinery/resources/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	config2 "github.com/siderolabs/talos/pkg/machinery/config/config"
+	"github.com/siderolabs/talos/pkg/machinery/config/types/v1alpha1"
+	"github.com/siderolabs/talos/pkg/machinery/proto"
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+)
+
+//go:generate deep-copy -type RegistriesSpecSpec -header-file ../../../../hack/boilerplate.txt -o deep_copy.generated.go .
+
+// RegistriesSpecType is type of RegistriesSpec resource.
+const RegistriesSpecType = resource.Type("Registries.net.talos.dev")
```
Comment (smira) [inline pkg/machinery/resources/registry/registry.go:24]: let's move this to `cri/` as it's the closest directory matching the purpose

File: internal/app/machined/pkg/controllers/files/cri_registry_config.go [hunk 2]
```diff
@@ -88,23 +89,23 @@ func (ctrl *CRIRegistryConfigController) Run(ctx context.Context, r controller.R
 		case <-r.EventCh():
 		}
 
-		cfg, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		cfg, err := safe.ReaderGetByID[*registry.RegistriesSpec](ctx, r, registry.RegistriesID)
```
Comment (smira) [inline internal/app/machined/pkg/controllers/files/cri_registry_config.go:92]: this can't work, as the `Inputs()` haven't been updated

**Review by smira — COMMENTED** [review]
the image/resolver (via PullImage) hasn't been updated

File: pkg/machinery/resources/registry/registry.go (hunk 1 above)
Comment (DmitriyMV) [reply pkg/machinery/resources/registry/registry.go]: Done.

File: internal/app/machined/pkg/controllers/files/cri_registry_config.go (hunk 2 above)
Comment (DmitriyMV) [reply internal/app/machined/pkg/controllers/files/cri_registry_config.go]: Fixed.

**Comment by DmitriyMV** [comment]
@smira I think all paths leading to `image` package are updated now.

File: hack/structprotogen/proto/proto.go [hunk 3]
```diff
@@ -396,6 +396,12 @@ func formatTypeName(fieldTypePkg string, fieldType string, declPkg string) (stri
 		return commoProto, "common.PEMEncodedCertificate"
 	case typeData{"github.com/siderolabs/talos/pkg/machinery/cel", "Expression"}:
 		return "google/api/expr/v1alpha1/checked.proto", "google.api.expr.v1alpha1.CheckedExpr"
+	case typeData{"github.com/siderolabs/talos/pkg/machinery/resources/cri", "RegistryMirrorConfig"}:
+		// This is a hack, but I (Dmitry) don't have enough patience to figure out why we don't support complex maps
+		return "resource/definitions/cri/registry.proto", "talos.resource.definitions.cri.RegistryMirrorConfig"
+	case typeData{"github.com/siderolabs/talos/pkg/machinery/resources/cri", "RegistryConfig"}:
+		// This is a hack, but I (Dmitry) don't have enough patience to figure out why we don't support complex maps
```
Comment (smira) [inline hack/structprotogen/proto/proto.go:403]: the author gave up :stuck_out_tongue:

File: pkg/machinery/resources/cri/registry.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/cosi-project/runtime/pkg/resource/typed"
+	"github.com/siderolabs/crypto/x509"
+
+	config2 "github.com/siderolabs/talos/pkg/machinery/config/config"
+	"github.com/siderolabs/talos/pkg/machinery/config/types/v1alpha1"
+	"github.com/siderolabs/talos/pkg/machinery/proto"
+)
+
+// RegistriesSpecType is type of RegistriesSpec resource.
+const RegistriesSpecType = resource.Type("Registries.net.talos.dev")
```
Comment (smira) [inline pkg/machinery/resources/cri/registry.go:20]: if we call it `registries`, we should name the constant `RegistriesType`, and cascading down.

I would probably name it `RegistriesConfig` though

File: pkg/machinery/resources/cri/registry.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+// RegistriesID is the singleton resource ID.
+const RegistriesID resource.ID = "registries"
+
+// RegistriesSpec resource holds info about container registries.
+type RegistriesSpec = typed.Resource[RegistriesSpecSpec, RegistriesSpecExtension]
```
Comment (smira) [inline pkg/machinery/resources/cri/registry.go:26]: add it to the `cri_test.go` in this folder to catch some issues quickly

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go [hunk 6]
```diff
@@ -1506,13 +1507,19 @@ func Upgrade(_ runtime.Sequence, data any) (runtime.TaskExecutionFunc, string) {
 
 		logger.Printf("performing upgrade via %q", in.GetImage())
 
+		rgs, err := safe.ReaderGetByID[*crires.RegistriesSpec](ctx, r.State().V1Alpha2().Resources(), crires.RegistriesID)
```
Comment (smira) [inline internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go:1510]: this certainly has a race condition (not here, but more on the install sequence)

a proper way would be to block for the config to be available

```suggestion
		rgs, err := safe.StateWaitFor[*crires.RegistriesSpec](ctx, r.State().V1Alpha2().Resources(), crires.RegistriesID, state.WithEventTypes(state.Created, state.Updated))
```

File: hack/structprotogen/proto/proto.go (hunk 3 above)
Comment (DmitriyMV) [reply hack/structprotogen/proto/proto.go:403]: Yeah...

File: pkg/machinery/resources/cri/registry.go (hunk 4 above)
Comment (DmitriyMV) [reply pkg/machinery/resources/cri/registry.go]: Done!

File: internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go (hunk 6 above)
Comment (DmitriyMV) [reply internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go]: Done!

**Comment by DmitriyMV** [comment]
Integration provision passes, so upgrades work. I'll remove the label to speed-up process.

**Comment by smira** [comment]
Missing bits:

* some simple unit-test for RegistriesConfig controller - it generate empty on no config, it generates something with config, it enables stuff with image cache config
* some unit-tests for registryd itself (might be super high-level, using httptest or lower level)

**Comment by frezbo** [comment]
> lots of other fixes here as we went through testing image cache

should we split those into a different PR, easier to backport if needed

**Comment by smira** [comment]
> > lots of other fixes here as we went through testing image cache
> 
> should we split those into a different PR, easier to backport if needed

this also works, probably `halt_if_installed` is the only important fix that can be backported

File: internal/app/machined/pkg/controllers/cri/registry.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/siderolabs/talos/pkg/machinery/resources/cri"
+)
+
+// ConfigController watches v1alpha1.Config, updates registry.RegistriesConfig.
+type ConfigController = transform.Controller[*config.MachineConfig, *cri.RegistriesConfig]
+
+// NewConfigController instanciates the config controller.
+//
+//nolint:gocyclo
+func NewConfigController() *ConfigController {
```
Comment (smira) [inline internal/app/machined/pkg/controllers/cri/registry.go:29]: should be probably `RegistriesConfigController` (named after the output resource)

Comment (DmitriyMV) [reply internal/app/machined/pkg/controllers/cri/registry.go]: Done

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go [hunk 8]
```diff
@@ -127,13 +128,11 @@ func (s *Server) checkSupported(feature runtime.ModeCapability) error {
 
 func (s *Server) checkControlplane(apiName string) error {
 	switch s.Controller.Runtime().Config().Machine().Type() { //nolint:exhaustive
```
Comment (smira) [inline internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go:130]: if touching this, better to refactor using `.IsControlplane()` method of the `machine.Type`

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go [hunk 9]
```diff
@@ -1228,19 +1227,19 @@ func (s *Server) Logs(req *machine.LogsRequest, l machine.MachineService_LogsSer
 
 	switch {
 	case req.Namespace == constants.SystemContainerdNamespace || req.Id == "kubelet":
-		var options []runtime.LogOption
+		var opts []runtime.LogOption
```
Comment (smira) [inline internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go:1230]: can we please skip in the future changing every line for unrelated changes, and close to beta release specifically. it makes more work for reviewers.

File: internal/app/machined/pkg/controllers/cri/registry.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (smira) [inline internal/app/machined/pkg/controllers/cri/registry.go:1]: the file should be `registries_config.go` to match controller name

File: internal/app/machined/pkg/controllers/cri/registry_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	crires "github.com/siderolabs/talos/pkg/machinery/resources/cri"
+)
+
+type ConfigSuite struct {
+	ctest.DefaultSuite
+}
+
+func (suite *ConfigSuite) TestRegistry() {
+	cfg := config.NewMachineConfig(container.NewV1Alpha1(&v1alpha1.Config{
```
Comment (smira) [inline internal/app/machined/pkg/controllers/cri/registry_test.go:28]: there's a missing case for missing machine config

File: internal/app/machined/pkg/system/services/registry/app/main.go [hunk 12]
```diff
@@ -38,9 +37,9 @@ func app() error {
 		return fmt.Errorf("failed to get user home directory: %w", err)
 	}
 
-	it := func(yield func(fs.StatFS) bool) {
+	it := func(yield func(string2 string) bool) {
```
Comment (smira) [inline internal/app/machined/pkg/system/services/registry/app/main.go:40]: `string2` ? or should be path?

**Review by smira — APPROVED** [review]

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go (hunk 8 above)
Comment (DmitriyMV) [reply internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go]: Dropped all irrelevant stuff.

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go (hunk 9 above)
Comment (DmitriyMV) [reply internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go]: Dropped all irrelevant stuff.

File: internal/app/machined/pkg/controllers/cri/registry.go (hunk 10 above)
Comment (DmitriyMV) [reply internal/app/machined/pkg/controllers/cri/registry.go]: Done.

File: internal/app/machined/pkg/controllers/cri/registry_test.go (hunk 11 above)
Comment (DmitriyMV) [reply internal/app/machined/pkg/controllers/cri/registry_test.go]: Added `TestRegistryNoMachineConfig`.

File: internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go (hunk 9 above)
Comment (DmitriyMV) [reply internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go]: Sorry! I thought you wanted me to drop this unrelated stuff.

**Pull request merged**